### DISCUSSION TITLE
Reworking audiobookshelf for use in a university course

### DISCUSSION THREAD

**Discussion by ruoti** [advplyr/audiobookshelf#1115] [General]
So, as background, I am a professor at the University of Tennessee teaching computer science. I'm looking to develop a web design class that I would start teaching probably a year and a half from now.

For this class, I'm thinking about having a semester-long project built around audiobookshelf. However, the codebase as its currently structured isn't ideal for teaching a class (to be expected as pedagogy is different from production). To prep for this class, I'm thinking about rewriting audiobookshelf using Nuxt 3 and typescript. While I'll try to stay true to the general look/feel/functionality of audiobookshelf, my first priority will be using various development paradigms that I want to instruct students on.

While my effort will likely never catch up to being feature complete with audiobookshelf (not necessary for my teaching needs), it may be possible that it will help if it is decided to migrate audiobookshelf to nuxt 3. In this vein, I'm curious about two questions:

1. Are there any things you wished you had done differently in the design of audiobookshelf?
2. What larger features have you been interested in integrating that might be too difficult with the current setup?

These are things I would love to look into when trying my refactoring. Looking at discussions, I see some topics could include theming or internationalization, but I'm sure there are many more.

I would be interested in any thoughts you are willing to share. If we're lucky, this effort won't just be useful for my students, but maybe this project down the road.

**Comment by advplyr** [comment]
This is a great course idea. I would be very interested in seeing what you put together if you go this route.

Most of the things I would do differently now aren't related to the frontend design. For example, abs is generating a static frontend before packaging for deployment. This makes it difficult to support using a subfolder path as mentioned here #385. The solutions to that all involve a significant rewrite.

Another thing I would do differently is use a sqlite database instead of the simple json database (this refactor is going to happen eventually). There are a lot more non-frontend things I would do different but I'll leave it there since it wouldn't be relevant to a web design course.

Here are a few things with the frontend

1. Theme customization + dark mode. With tailwind implementing dark mode is not bad but it would have been easier had all of the components been built with this in mind.
2. Localization. This won't be difficult to add in now but could have saved some time had I just started with it.
3. Library selected in the appbar. I'm on the fence about this one but there have been some challenges by having the selected library be global like it is now. Other media servers like Plex & Jellyfin have a home page that combines all libraries and a search input that searches across all libraries. Putting the library selection at the top has its benefits like allowing us to have a siderail nav with Library/Series/Collections/etc., but makes it difficult design-wise to do global searching and a combined home page. Relevant issue #556
4. Podcasts. I'm not sure if your version would include podcasts but the design of podcasts in abs still feels tacked on. It's not clear to new users how to add a new podcast and the podcast landing page is not performant if you have thousands of episodes. Relevant issue #960
5. Playlists & player queue. Collections are meant to be for all users but we should also have playlists that are just for the user. It would have saved time had the audio player been built with a queue in mind. Player queue and playlists would be similar to Youtube (Plex does the same I believe) where you can add media to the currently playing queue then save that queue into a playlist if you wanted.
6. Multiple filters + searching in the sort/filter dropdowns. This is just a nice to have feature I'd like to put in. Users may have thousands of narrators for example and to filter by one would need to scroll through a long list, so having a small search box at the top will be nice. Relevant issue #1080
7. More robust user permission system. This is mostly backend & data model but having the option to restrict a user to a specific genre for example has been requested a few times. Relevant issue #968

I may think of some more to throw in here later. Let me know if you want me to expand on any of those.

You can also email me advplyr@protonmail.com, I'd love to be of assistance if I can with your work.

**Reply by ddubnyk** [reply]
I see you mention here about eventually going to move to a sqlite database.  When this happens do you also plan to allow using an external database of our own?  I run my own Postgres database server for a lot of other service, as it makes backing things up really simple along with more options and performance enhancements.

**Reply by advplyr** [reply]
Maybe but haven't really thought about it. Moving to a relational db hasn't even started yet so there is a long way to go before that is a possibility.

**Comment by ruoti** [comment]
Great suggestions on the front end. This web design class will likely cover both frontend and backend development to some extent, so I'm also curious what design decisions you would have done differently on the backend. Nuxt 3 actually provides a backend server as well as front end components.
